### ISSUE TITLE
Hard: migrate from owners.json and voxels.json to Postgres database

### ISSUE THREAD

**Issue by kf106** [kf106/orthoverse-prismarine#55]
Information scraped from the blockchain should be stored in a Postgres database and not in some JSON files that cannot be dynamically updated as the Orthoverse voxel game runs. We will also track Orthoverse-avatars database integration here to keep everything in one place (as orthoverse-prismarine appears to be the main repo for tracking issues, unless there is a way to have a co-mingled issue tracking system on Github).

**Comment by kf106** [comment]
* One Postgres database running on orthoverse.land to track all dynamic data
* The orthoverse-offline to write all scraped token data to that database
* Orthoverse-website to read relevant data from the database
* Orthoverse-prismarine to get ownership and location data from the database, from the lands table
* Orthoverse-avatars to use a second table, avatars, to store character image data from user's designing stuff

I suggest that we begin with orthoverse-prismarine and orthoverse-website being upgraded and functioning fully before we start on orthoverse-avatars

**Comment by kf106** [comment]
This is now tracked in a private google doc.

**Comment by kf106** [comment]
orthoverse.io has an api that gets you all the details you need for a given land.

But the world generator function is synchronous, so adding an api call causes async problems
Furthermore, the world generator function generates the world by chunks (16x16x256 columns), not lands (96x96x256 blocks), so the simple solution ends up making far more API calls than necessary.

The design for integrating the land database therefore needs to be rethought.

The mca files cannot currently be pregenerated, because it is done asynchronous per chunk, and this leads to a memory overflow if an attempt to generate the entire world is made.

Initial idea:  rely on standard world generator, use a worker to scan the database for changes, and save them to land-save files (using 0 at the default slot for this). Use the load/save functionality to pull in local changes.

**Comment by kf106** [comment]
To start with, I have moved from using `const voxel = require('../../map-data/voxel.json')` in numerous files to loading the json map into `serv.voxel.data`and providing the helper methods `serv.voxel.load()` and `serv.voxel.save()` which means we can dynamically update the stored map data during runtime.

The flying-squid references have all been updated, and it all appears to work.

- [x] Update prismarine-web-client to use ethereum channel queries to flying-squid rather than also loading with `const voxel = require('../../map-data/voxel.json')`

There is a problem with the land saves, in that the position of each chunk is hard-coded into the chunk save file.

- [ ] check with `voxel.json`, or rather `serv.voxel.json`, where the chunk is supposed to go, and update the absolute position of the chunk accordingly before loading it.
- [x] ~~write a helper program to scrape the database and construct and write out `oxel.json`, which is `voxel.json` but with object properties that are readable rather than having to remember what voxel[0] and voxel[4] actually mean.~~ 
- [x] ~~Then move from using `voxel.json` to `oxel.json`. Then `voxel.json` will just be for orthoverse-offline and keeping the database up to date and there will be less confusion. And oxel means "orthoverse voxels".~~
- [x] flying-squid should regularly write out the current state of the `serv.oxel.data` object to keep the world up to date
- [ ] `serv.oxel.data` and `oxel.json` should store the selected save version by a player/owner of their land, and should use that in preference to default orthogen.js generated lands
- [ ] Perhaps we can use orthogen to create 0.land and 8.land to store the default state of the land. Pressing 0 could always reload that.

**Comment by kf106** [comment]
Note that resolving all of this may magically fix https://github.com/Orthocorp/orthoverse-prismarine/issues/75 which I think is an edge case of saving when the whole land hasn't been loaded yet.

**Comment by kf106** [comment]
The new API call at https://orthoverse.io/api/land/generate/voxel returns the equivalent of voxels.json in less than a second so the need for a file has disappeared.

**Comment by kf106** [comment]
Forget the checklist above.

Webclient now makes an API call to work out who owns the land being entered into. This gets the correct name, shield and crown. The webclient gets the world details through chunks sent over a communication directly to the server, including updates to land, so that's handled.

Flying-Squid is where the problem now lies. The world mca files reflect what the generator function deduced, and then overwrites from land loads.

Workflow so far:
- on starting the server, try to load from doxels.json file into serv.voxel.data. If there is no file, load from the database, then save to file. I have called this file doxels.json to avoid accidentally picking up an old voxels.json from a shields.sh copying activity or something like that. Pretend that it stands for "voxels.json plus datetime", because it also contains a "timestamp" key, which is the time at which it was created from the database.
- this means that serv.voxels.data['timestamp'] contains the time at which the data was accurate.
- now note that we can query the database at any time for any changes to individual lands: ownership change, or new tokens. Tokens may also move around when the 5000 reveal second wave arrives, so I need to keep that in mind
- the diff tells us which tokens have changed owner. This is picked up instantly by webclient - flying-squid needs it to ensure that new owners of tokens can dig and build on their land, which probably also explains why my son's token gives him a build bar but not the ability to build
- changes to the database currently only happen after a shields.sh scraping run, every four hours, so orthoverse-prismarine only needs to scrape for changes infrequently.

**Comment by kf106** [comment]
Now we need to keep track of what coordinates (6x6 sets of chunks that make up a land) have been visited and loaded, and what was loaded into them. This could be: a default land from orthogen, a saved and then loaded land, or open sea because there was no land for that coordinate.

This should be stored in a simple object, or possibly in serv.voxel.data - although that would mean that we can't overwrite serv.voxel.data, we have to copy across instead to prevent overwriting keys in co-ordinate objects.

Let's start with a second object, serv.voxel.status - key is "x:y", values are:
```
tokenId <address>, // the 0xabc...123 for the token
inMca? <bool>, // tells us if the land has been visited and generated into the world file
saveSlotUsed <int>, // tells us which save slot is currently in use
realm <string: futuristic | fantasy > // tells us whether to use data from land-saves fantasy or futuristic folder
```

saveSlotUsed can be used to indicate whether save slot 1-7 is used, or 0 to indicate it's orthogen generated.

The is a problem if a land is built on, but it's not saved, so perhaps lands should regularly be saved to files to hold that information, or perhaps we're just mean to level 0 builders.

Then we regularly run loadDiff and loop through all the changes, handling the changes.

Cases:
- a new land has been revealed, but it's not in the mca: we don't have to worry, as it will be picked up first time someone visits
- a new land has been revealed, and the open sea that was there is in the mca: we need to overwrite that section of the mca with details obtained from orthogen
- an old land has leveled up and it's not in the mca: not a problem
- an old land has leveled up and it is in the mca and has never been saved: we need to overwrite that section with a new run of orthogen
- an old land has leveled up and is in the mca and has been saved: the saved data is fine

**Comment by kf106** [comment]
Further cases:
- two lands have swapped positions. Then things get complicated, because land save files include the coordinates of the chunks, and they'll need changing. So that involves piping them through prismarine-chunk. Yuck.

Further problems may occur if only part of a land has been loaded - I'll need to think about that more. Perhaps I can ensure that when a new orthogen call is needed, it does so for the entire land, so we don't ever end up with partial lands...

**Comment by kf106** [comment]
I have moved loadLand and saveLand from ethereum.js to index.js as part of the main mcServer class, thus exposing them to other areas of the server. They should be usable to reload saved lands by the server based on data from the serv.voxels.data and serv.voxels.data, and loads of other clever stuff.

**Comment by kf106** [comment]
The basic system now works - and lands are saved under their name because I got tired of looking up 40 character address keys to find a land (strangely people like to call their land things like Unrimmark rather than 0xcefd4335...457fe).

**Comment by kf106** [comment]
Fixed by PR #79
